File: AppLovin MAX Demo App - Java/app/src/main/java/com/applovin/enterprise/apps/demoapp/ui/BaseAdActivity.java
package com.applovin.enterprise.apps.demoapp.ui;

import com.applovin.enterprise.apps.demoapp.R;

import java.util.ArrayList;
import java.util.List;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

/**
 * Base for activities showing ads.
 * <p>
 * Created by Harry Arakkal on 2019-10-21.
 */
public abstract class BaseAdActivity
        extends AppCompatActivity
{
    private       CallbacksRecyclerViewAdapter callbacksAdapter;
    private final List<String>                 callbacks = new ArrayList<>();

    /**
     * Setup callbacks RecyclerView adapter and appearance.
     */
    protected void setupCallbacksRecyclerView()
    {
        callbacksAdapter = new CallbacksRecyclerViewAdapter( callbacks, this );
        LinearLayoutManager manager = new LinearLayoutManager( this );
        DividerItemDecoration decoration = new DividerItemDecoration( this, manager.getOrientation() );

        RecyclerView callbacksRecyclerView = findViewById( R.id.callbacks_recycler_view );
        callbacksRecyclerView.setHasFixedSize( true );
        callbacksRecyclerView.setAdapter( callbacksAdapter );
        callbacksRecyclerView.setLayoutManager( manager );
        callbacksRecyclerView.addItemDecoration( decoration );
    }

    /**
     * Log ad callbacks in the RecyclerView.
     * Uses the name of the function that calls this one in the log.
     */
    protected void logCallback()
    {
        String callbackName = new Throwable().getStackTrace()[1].getMethodName();
        callbacks.add( callbackName );

        callbacksAdapter.notifyItemInserted( callbacks.size() - 1 );
    }

    protected void logAnonymousCallback()
    {
        String callbackName = new Throwable().getStackTrace()[2].getMethodName();
        callbacks.add( callbackName );

        callbacksAdapter.notifyItemInserted( callbacks.size() - 1 );
    }
}

File: AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ui/BaseAdActivity.kt
package com.applovin.enterprise.apps.demoapp.ui

import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.applovin.enterprise.apps.demoapp.R

/**
 * Base for activities showing ads.
 * <p>
 * Created by Harry Arakkal on 2019-10-18.
 */
abstract class BaseAdActivity : AppCompatActivity()
{
    private lateinit var callbacksAdapter: CallbacksRecyclerViewAdapter
    private val callbacks: MutableList<String> = mutableListOf()

    /**
     * Setup callbacks RecyclerView adapter and appearance.
     */
    protected fun setupCallbacksRecyclerView()
    {
        callbacksAdapter = CallbacksRecyclerViewAdapter(callbacks, this)
        val manager = LinearLayoutManager(this)
        val decoration = DividerItemDecoration(this, manager.orientation)

        val callbacksRecyclerView = findViewById<RecyclerView>(R.id.callbacksRecyclerView)
        callbacksRecyclerView.apply {
            setHasFixedSize(true)
            adapter = callbacksAdapter
            layoutManager = manager
            addItemDecoration(decoration)
        }
    }

    /**
     * Log ad callbacks in the RecyclerView.
     * Uses the name of the function that calls this one in the log.
     */
    protected fun logCallback()
    {
        val callbackName = Throwable().stackTrace[1].methodName
        callbacks.add(callbackName)

        callbacksAdapter.notifyItemInserted(callbacks.lastIndex)
    }

    protected fun logAnonymousCallback()
    {
        val callbackName = Throwable().stackTrace[2].methodName
        callbacks.add(callbackName)

        callbacksAdapter.notifyItemInserted(callbacks.lastIndex)
    }
}
